### ISSUE TITLE
App Crashes with Library not loaded: @rpath/WebRTC.framework/WebRTC on Device

### ISSUE THREAD

**Issue by mohit-kapadia-at-lirisoft** [livekit/client-sdk-swift#371]
**Describe the bug**
App crashes while running on a physical device. Added a sample project(TestLiveKitIntegration.zip) to run and reproduce.

**SDK Version**
1.1.6, Adding the sdk using SPM under a sub project. Added screenshot. Even tried the latest version 2.0.6 

**iOS/macOS Version**
17.3(iPhone 13 Pro Max)

**Steps to Reproduce**
- Created a sub module(sub project) in a Xcode project
- Added live-kit as the package to the sub project.

**Expected behavior**
App Crashes on launch on device. It should not crash.

**Screenshots**
Addeed Screenshots

**Logs**
dyld[579]: Library not loaded: @rpath/WebRTC.framework/WebRTC
  Referenced from: <0918C798-1BD0-368A-AAFD-3861F4093113> /private/var/containers/Bundle/Application/A0EED870-68BD-494C-9417-705F03C8BEC8/TestLiveKitIntegration.app/Frameworks/SubProjectFramework.framework/SubProjectFramework
  Reason: tried: '/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/preboot/Cryptexes/OS/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/var/containers/Bundle/Application/A0EED870-68BD-494C-9417-705F03C8BEC8/TestLiveKitIntegration.app/Frameworks/WebRTC.framework/WebRTC' (no such file), '/private/var/containers/Bundle/Application/A0EED870-68BD-494C-9417-705F03C8BEC8/TestLiveKitIntegration.app/Frameworks/SubProjectFramework.framework/Frameworks/WebRTC.framework/WebRTC' (no such file), '/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/preboot/Cryptexes/OS/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/var/containers/Bundle/Application/A0EED870-68BD-494C-9417-705F03C8BEC8/TestLiveKitIntegration.app/Frameworks/WebRTC.framework/WebRTC' (no such file), '/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/preboot/Cryptexes/OS/Users/newuser/Library/Developer/Xcode/DerivedData/TestLiveKitIntegration-cqtlgdcnzeuwukgnvurxufzyqtdc/Build/Products/Debug-iphoneos/PackageFrameworks/WebRTC.framework/WebRTC' (no such file), '/private/var/containers/Bundle/Application/A0EED870-68BD-494C-9417-705F03C8BEC8/TestLiveKitIntegration.app/Frameworks/WebRTC.framework/WebRTC' (no such file)

<img width="1789" alt="Screenshot 2024-05-03 at 7 25 17 PM" src="https://github.com/livekit/client-sdk-swift/assets/81983419/b8a1c525-36c3-41a7-a701-f1e65d8f56a9">

[TestLiveKitIntegration.zip](https://github.com/livekit/client-sdk-swift/files/15200986/TestLiveKitIntegration.zip)

**Comment by davidzhao** [comment]
Please remove WebRTC as a dependency, we no longer reference it in v2

**Comment by mohit-kapadia-at-lirisoft** [comment]
We have not added it explicitly, LiveKit is adding it as its own dependency. Pls check this screenshot.

<img width="1792" alt="Screenshot 2024-05-05 at 11 14 45 AM" src="https://github.com/livekit/client-sdk-swift/assets/81983419/635d4dae-d042-49ab-9987-77bd83c90891">

**Comment by davidzhao** [comment]
you'll want to point it at 2.0, which has the [following deps](https://github.com/livekit/client-sdk-swift/blob/main/Package.swift)

```
    dependencies: [
        // LK-Prefixed Dynamic WebRTC XCFramework
        .package(url: "https://github.com/livekit/webrtc-xcframework.git", exact: "114.5735.14"),
        .package(url: "https://github.com/apple/swift-protobuf.git", from: "1.26.0"),
        .package(url: "https://github.com/apple/swift-log.git", from: "1.5.4"),
        // Only used for DocC generation
        .package(url: "https://github.com/apple/swift-docc-plugin.git", from: "1.3.0"),
        // Only used for Testing
        .package(url: "https://github.com/vapor/jwt-kit.git", from: "4.13.4"),
    ],
```

**Comment by mohit-kapadia-at-lirisoft** [comment]
I see the same in latest version. Pls check the screenshot, or if you can try the sample that I have added in earlier comment, you can reproduce it. Thanks for responding so quickly.

<img width="1789" alt="Screenshot 2024-05-06 at 11 28 20 AM" src="https://github.com/livekit/client-sdk-swift/assets/81983419/214597ca-4cf0-4564-b829-a9fd095bd47e">

**Comment by codin-play** [comment]
I think I may have the same issue as you @mohit-kapadia-at-lirisoft . Did you determine a way around this?

**Comment by hiroshihorie** [comment]
Hello, sorry for the late reply.
I've tried your project with v2.0.10 and it seems to work. (Xcode 15.4)

![Screenshot 2024-06-18 23 20 30](https://github.com/livekit/client-sdk-swift/assets/548776/891a2c8a-b39f-4fac-b3e9-89a9dfc02d3c)

Can you try `Clean Build Folder` and re-running ?
Which Xcode version are you using ?

**Comment by janhuachu** [comment]
We are still having problem with this. We built a xcframework which include LiveKit package dependencies.  Then our app include that xcframework. The app runs fine on the simulator but crashed on the real device.

We are using LiveKit swift package version 2.12.1. Any further investigation to resolve this issue?

**Comment by xianshijing-lk** [comment]
Thanks @janhuachu for reporting it, I just reopened the ticket and will come back to you shortly.

**Comment by pblazej** [comment]
@janhuachu sorry for the inconvenience, did you have a chance to try this method? https://github.com/livekit/client-sdk-swift/issues/371#issuecomment-2176239845

**Comment by janhuachu** [comment]
> [@janhuachu](https://github.com/janhuachu) sorry for the inconvenience, did you have a chance to try this method? [#371 (comment)](https://github.com/livekit/client-sdk-swift/issues/371#issuecomment-2176239845)

@pblazej  
For a workaround, we are either downloading the corresponding LiveKitWebRTC.xcframework from LiveKit git hub, or include the same LiveKit swift package in the app.

**Comment by pblazej** [comment]
> downloading the corresponding LiveKitWebRTC.xcframework

@janhuachu then you copy the frameworks manually/with a script, right?

> or include the same LiveKit swift package in the app

This is a valid solution - just make sure the version matches.

Is it working on device? I want to confirm is it a hard **blocker** or "inconvenience" in your current setup. I tested your example project from above ⏫ and I'm unable to reproduce the crash, after applying the above workarounds.

**Comment by pblazej** [comment]
Another working solution is to add a following script to the app build phases:

```sh
# Workaround for SPM bug: dynamic xcframework dependencies from
# sub-project frameworks are not embedded in the app bundle.
# https://github.com/apple/swift-package-manager/issues/6069
set -e
FW="LiveKitWebRTC"
SRC="${BUILT_PRODUCTS_DIR}/${FW}.framework"
DEST="${BUILT_PRODUCTS_DIR}/${FRAMEWORKS_FOLDER_PATH}"
if [ ! -d "${SRC}" ]; then
    echo "warning: ${FW}.framework not found at ${SRC}, skipping"
    exit 0
fi
mkdir -p "${DEST}"
ditto "${SRC}" "${DEST}/${FW}.framework"
if [ "${CODE_SIGNING_REQUIRED:-NO}" = "YES" ] && [ -n "${EXPANDED_CODE_SIGN_IDENTITY:-}" ]; then
    codesign --force --sign "${EXPANDED_CODE_SIGN_IDENTITY}" --preserve-metadata=identifier,entitlements --timestamp=none "${DEST}/${FW}.framework"
fi
echo "Embedded ${FW}.framework into app bundle"
```

with the following config:

<img width="791" height="457" alt="Image" src="https://github.com/user-attachments/assets/46257ff1-55d1-45ea-b570-bb4205db9f18" />

⚠️ it may require `ENABLE_USER_SCRIPT_SANDBOXING = NO;` setting at the target level.

Probably the safest one is still:

> or include the same LiveKit swift package in the app

**Comment by janhuachu** [comment]
I agree that the workaround is fine if this is a swift package handling problem as [#371 comment](https://github.com/livekit/client-sdk-swift/issues/371#issuecomment-2176239845)

**Comment by pblazej** [comment]
Closing that as it exact setup requires SPM fixes.
